myia/unity.py

```diff
@@ -0,0 +1,503 @@
+            filter: FilterT[T] = None)-> Var[T]:
+        """Create a variable for unification purposes.
+
+        Arguments:
+            tag: An identifier for the variable. Two variables with the
+                same filter and identifier will return the same object.
+            filter: A predicate, or a set of values the variable is
+                allowed to take.
+        """
+        tag = self._get_tag(tag)
+        if callable(filter):
+            return self._get_var(FilterVar, (tag, filter))
+        elif filter:
+            return self._get_var(RestrictedVar, (tag, filter))
+        else:
+            return self._get_var(Var, (tag,))
+
+    def svar(self, tag: str = None) -> SVar[T]:
+        """Create an SVar (can match 0 or more items)."""
+        return self._get_var(SVar, (self._get_tag(tag),))  # type: ignore
+
+    def uvar(self, values: Iterable[Value[T]], tag: str = None) -> UnionVar[T]:
+        """Create a UnionVar (represents multiple possibilities)."""
+        return self._get_var(UnionVar,  # type: ignore
+                             (self._get_tag(tag), values))
+
+    class VisitError(Exception):
+        """Report unvisitable object."""
+
+    def visit(self, fn: Callable[[Value[T]], Value[T]],
+              value: Value[T]) -> Value[T]:
+        """Apply `fn`to each element of `value` and return the result."""
+        raise self.VisitError
+
+    def clone(self, v: Value[T], copy_map: Dict[Any, Any] = None) -> Value[T]:
+        """Return a copy of a templated type structure.
+
+        Type are passed through without modification, variables are
+        duplicated with a new id.
+
+        This preserves relationships between variables like this::
+
+            clone(Tuple(v1, v1)) -> Tuple(v2, v2)
+
+        Arguments:
+            v: expression
+            copy_map: Dictionnary of variable mapping
+
+        """
+        if copy_map is None:
+            copy_map = {}
+        return self._clone(v, copy_map)
+
+    def _clone(self, v: Value[T], copy_map: Dict[Any, Any]) -> Value[T]:
+        if v in copy_map:
+            return copy_map[v]
+
+        if isinstance(v, Var):
+            if isinstance(v, SVar):
+                copy_map[v] = self.svar()
+            elif isinstance(v, UnionVar):
+                copy_map[v] = self.uvar(self._clone(v, copy_map)
+                                        for v in v.values)
+            else:
+                filter: FilterT[T] = None
+                if isinstance(v, FilterVar):
+                    filter = v.filter
+                elif isinstance(v, RestrictedVar):
+                    filter = v.legal_values
+                copy_map[v] = self.var(filter=filter)
+
+        elif isinstance(v, Seq):
+            copy_map[v] = Seq(self._clone(val, copy_map) for val in v)
+
+        else:
+            try:
+                return self.visit(lambda v: self._clone(v, copy_map), v)
+            except self.VisitError:
+                return v
```

Comment:
Do we want to ignore this error?

Reply:
Yes, we assume it's a non-container value (like Bool() for types).